File: backend/src/main/java/com/grepp/smartwatcha/app/controller/api/admin/upcoming/UpcomingSyncApiController.java
package com.grepp.smartwatcha.app.controller.api.admin.upcoming;

import com.grepp.smartwatcha.app.controller.web.admin.movie.upcoming.UpcomingMovieSync;
import com.grepp.smartwatcha.app.model.admin.movie.upcoming.dto.UpcomingMovieSyncDto;
import com.grepp.smartwatcha.infra.response.ApiResponse;
import com.grepp.smartwatcha.infra.response.ResponseCode;
import java.util.List;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/*
 * 공개 예정작 동기화를 위한 API 컨트롤러
 * TMDB 에서 공개 예정작 정보를 가져와 DB에 저장하는 동기화 작업을 수행
 */
@Slf4j
@RestController
@RequiredArgsConstructor
@RequestMapping("/admin/movies/upcoming")
public class UpcomingSyncApiController {

  private final UpcomingMovieSync scheduler;

  @PostMapping("/sync")
  public ResponseEntity<ApiResponse<UpcomingMovieSyncDto>> syncUpcomingManually() {
    try {
      log.info("Starting manual upcoming movies sync...");
      UpcomingMovieSyncDto result = scheduler.syncAllUpcomingMovies();
      log.info("Manual upcoming movies sync completed successfully");

      return ResponseEntity.ok(ApiResponse.success(result));

    } catch (Exception e) {
      log.error("Failed to sync upcoming movies: {}", e.getMessage(), e);

      UpcomingMovieSyncDto errorDto = UpcomingMovieSyncDto.builder()
          .total(0)
          .success(0)
          .failed(1)
          .skipped(0)
          .enrichFailed(0)
          .skippedIds(List.of())
          .skippedReasons(List.of("예외: " + e.getMessage()))
          .failedIds(List.of())
          .build();

      return ResponseEntity
          .status(ResponseCode.INTERNAL_SERVER_ERROR.status())
          .body(ApiResponse.error(ResponseCode.INTERNAL_SERVER_ERROR, errorDto));
    }
  }
}

File: userserviceserver/src/main/kotlin/com/grepp/smartwatcha/app/controller/api/upcomingmovie/UpcomingSyncApiController.kt
package com.grepp.smartwatcha.app.controller.api.upcomingmovie

import com.grepp.smartwatcha.app.model.upcomingmovie.dto.UpcomingMovieDto
import com.grepp.smartwatcha.app.model.upcomingmovie.service.UpcomingMovieSyncService
import com.grepp.smartwatcha.app.model.upcomingmovie.service.UpcomingSyncToSpringService
import kotlinx.coroutines.runBlocking
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/admin/movies/upcoming")
class UpcomingSyncApiController(
    private val syncService: UpcomingMovieSyncService,
    private val syncToSpringService: UpcomingSyncToSpringService
) {

    @Value("\${internal.token}")
    private lateinit var internalToken: String

    private val log = LoggerFactory.getLogger(UpcomingSyncApiController::class.java)

    // Kotlin → Spring 서버로 enrich된 DTO 전송용 (Spring 서버가 JSON List로 받음)
    @PostMapping("/sync")
    fun syncUpcomingMovies(
        @RequestHeader("x-internal-token", required = false) token: String?
    ): ResponseEntity<List<UpcomingMovieDto>> {
        if (token != internalToken) {
            log.warn("❌ 인증 실패: 토큰 불일치")
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build()
        }

        val result: List<UpcomingMovieDto> = runBlocking {
            syncService.syncAllAndReturnDtoList()
        }

        log.info("✅ Kotlin 서버에서 공개 예정작 동기화 완료 (총 ${result.size}건)")
        return ResponseEntity.ok(result)
    }

    // enrich 완료 후 Spring 서버에 전송 (RestTemplate 기반 호출)
    @GetMapping("/sync-and-send")
    suspend fun syncAndSendToSpring(): ResponseEntity<String> {
        val result = syncService.syncAll()
        syncToSpringService.sendToSpring(result.dtoList)
        return ResponseEntity.ok("✅ enrich + Spring 전송 완료")
    }
}
